File: src/2025-02-23-deepclaude-try/python/exercises/practice/connect/connect.py

class ConnectGame:
    def __init__(self, board):
        # Split board into rows and parse each row into cells
        self.grid = []
        if isinstance(board, str):
            rows = board.strip().split('\n')
        else:
            rows = board
            
        for row in rows:
            # Split row into cells, filtering out empty strings from spaces
            cells = [cell for cell in row.split(' ') if cell]
            self.grid.append(cells)
    
    def get_winner(self):
        # Check O's win condition (top to bottom)
        if self._check_o_win():
            return 'O'
        # Check X's win condition (left to right) 
        if self._check_x_win():
            return 'X'
        return ''
    
    def _check_o_win(self):
        if not self.grid:
            return False
            
        # Check from first row to last row
        start_row = 0
        target_row = len(self.grid) - 1
        
        # Start from any O in first row
        for col in range(len(self.grid[0])):
            if self.grid[0][col] == 'O':
                if self._can_reach_bottom(start_row, col):
                    return True
        return False
        
    def _check_x_win(self):
        if not self.grid:
            return False
            
        # Check each row for left-to-right connection
        for row in range(len(self.grid)):
            # Start from leftmost X
            if len(self.grid[row]) > 0 and self.grid[row][0] == 'X':
                if self._can_reach_right(row, 0):
                    return True
        return False
    
    def _can_reach_bottom(self, row, col):
        visited = set()
        queue = [(row, col)]
        visited.add((row, col))
        
        while queue:
            current_row, current_col = queue.pop(0)
            
            # Check if reached bottom row
            if current_row == len(self.grid) - 1:
                return True
                
            # Check all neighbors
            for next_row, next_col in self._get_neighbors(current_row, current_col):
                if (next_row, next_col) not in visited and self.grid[next_row][next_col] == 'O':
                    queue.append((next_row, next_col))
                    visited.add((next_row, next_col))
        return False
        
    def _can_reach_right(self, row, col):
        visited = set()
        queue = [(row, col)]
        visited.add((row, col))
        
        while queue:
            current_row, current_col = queue.pop(0)
            
            # Check if reached rightmost position in any row
            if current_col == len(self.grid[current_row]) - 1:
                return True
                
            # Check all neighbors
            for next_row, next_col in self._get_neighbors(current_row, current_col):
                if (next_row, next_col) not in visited and self.grid[next_row][next_col] == 'X':
                    queue.append((next_row, next_col))
                    visited.add((next_row, next_col))
        return False
    
    def _get_neighbors(self, row, col):
        # Returns valid neighboring coordinates in hexagonal grid
        neighbors = []
        # Check all 6 possible neighbors in hexagonal grid
        directions = [
            (-1, -1), (-1, 0),  # Upper neighbors
            (0, -1), (0, 1),    # Same row neighbors
            (1, -1), (1, 0)     # Lower neighbors
        ]
        
        for dr, dc in directions:
            new_row = row + dr
            new_col = col + dc
            if (0 <= new_row < len(self.grid) and 
                0 <= new_col < len(self.grid[new_row])):
                neighbors.append((new_row, new_col))
        return neighbors

File: src/2025-02-24-deepclaude-rust/go/exercises/practice/connect/connect.go
package connect

import (
	"strings"
)

func ResultOf(lines []string) (string, error) {
	if len(lines) == 0 {
		return "", nil
	}

	// Process grid by removing all spaces
	grid := make([][]rune, len(lines))
	for i, line := range lines {
		grid[i] = []rune(strings.ReplaceAll(line, " ", ""))
	}

	// Check O's top-to-bottom connection
	oVisited := make([][]bool, len(grid))
	for i := range oVisited {
		oVisited[i] = make([]bool, len(grid[i]))
	}

	var queue [][2]int
	directions := [][2]int{{-1, 0}, {-1, 1}, {0, -1}, {0, 1}, {1, 0}, {1, -1}}

	// Initialize O's BFS from top row
	for c := 0; c < len(grid[0]); c++ {
		if grid[0][c] == 'O' {
			queue = append(queue, [2]int{0, c})
			oVisited[0][c] = true
		}
	}

	foundO := false
	for len(queue) > 0 {
		cell := queue[0]
		queue = queue[1:]
		r, c := cell[0], cell[1]

		if r == len(grid)-1 {
			foundO = true
			break
		}

		for _, dir := range directions {
			nr, nc := r+dir[0], c+dir[1]
			if nr >= 0 && nr < len(grid) && nc >= 0 && nc < len(grid[nr]) {
				if grid[nr][nc] == 'O' && !oVisited[nr][nc] {
					oVisited[nr][nc] = true
					queue = append(queue, [2]int{nr, nc})
				}
			}
		}
	}

	if foundO {
		return "O", nil
	}

	// Check X's left-to-right connection
	xVisited := make([][]bool, len(grid))
	for i := range xVisited {
		xVisited[i] = make([]bool, len(grid[i]))
	}

	queue = queue[:0] // Reset queue

	// Initialize X's BFS from leftmost column
	for r := 0; r < len(grid); r++ {
		if len(grid[r]) > 0 && grid[r][0] == 'X' {
			queue = append(queue, [2]int{r, 0})
			xVisited[r][0] = true
		}
	}

	foundX := false
	for len(queue) > 0 {
		cell := queue[0]
		queue = queue[1:]
		r, c := cell[0], cell[1]

		if c == len(grid[r])-1 {
			foundX = true
			break
		}

		for _, dir := range directions {
			nr, nc := r+dir[0], c+dir[1]
			if nr >= 0 && nr < len(grid) && nc >= 0 && nc < len(grid[nr]) {
				if grid[nr][nc] == 'X' && !xVisited[nr][nc] {
					xVisited[nr][nc] = true
					queue = append(queue, [2]int{nr, nc})
				}
			}
		}
	}

	if foundX {
		return "X", nil
	}

	return "", nil
}
